### PR TITLE
Apply upgrades after state at block

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Continues to fix other tracer APIs suffering from similar issue as in #976

Closes: #992 

## How this works

Calls `ApplyUpgrades` after `StateAtBlock` if necessary.  

## How this was tested

Added unit tests.

## How is this documented

No need

### CODE REVIEW

File: eth/tracers/api.go [hunk 1]
```diff
@@ -957,7 +973,25 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
...
+		if config.BlockOverrides != nil && config.BlockOverrides.Time != nil {
+			modifiedTime := uint64(*config.BlockOverrides.Time)
+			if modifiedTime > originalTime {
+				doApplyUpgrades = true
+			}
+		}
+
 		config.BlockOverrides.Apply(&vmctx)
+		if doApplyUpgrades {
+			// Apply upgrades here as if the block was mined at the modified time.
```
Comment (darioush) [inline eth/tracers/api.go:989]:
```suggestion
			// Apply all relevant upgrades from [originalTime] to the block time set in the override.
```

File: eth/tracers/api.go [hunk 2]
```diff
@@ -957,7 +973,25 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
...
+		originalTime := block.Time()
+		if config.BlockOverrides != nil && config.BlockOverrides.Time != nil {
+			modifiedTime := uint64(*config.BlockOverrides.Time)
+			if modifiedTime > originalTime {
+				doApplyUpgrades = true
+			}
+		}
+
 		config.BlockOverrides.Apply(&vmctx)
+		if doApplyUpgrades {
```
Comment (darioush) [inline eth/tracers/api.go:988]: do we need the `if` statement?

File: eth/tracers/api_test.go [hunk 3]
```diff
@@ -1026,3 +1034,288 @@ func TestTraceBlockPrecompileActivation(t *testing.T) {
 		}
 	}
 }
+
+func TestTraceTransactionPrecompileActivation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply eth/tracers/api_test.go:1038]: These tests can be incorporated directly into the original tests (`TestTraceTransaction`). We can put those activations in original tests to reduce the diff. @darioush also asked to move those tests to a separate file https://github.com/ava-labs/subnet-evm/pull/999#discussion_r1396106739. @darioush do you think preserving upstream compatibility is better than reducing diffs?

[the remark this answers is not in the document]
Comment (ceyonur) [reply eth/tracers/api_test.go:1038]: replied https://github.com/ava-labs/subnet-evm/pull/999#discussion_r1397297357

File: eth/tracers/api.go (hunk 2 above)
Comment (ceyonur) [reply eth/tracers/api.go:988]: certainly not.

File: eth/state_accessor.go [hunk 4]
```diff
@@ -215,6 +215,13 @@ func (eth *Ethereum) stateAtTransaction(ctx context.Context, block *types.Block,
 	if err != nil {
 		return nil, vm.BlockContext{}, nil, nil, err
 	}
+
+	// Apply upgrades to the parent state
+	err = core.ApplyUpgrades(eth.blockchain.Config(), &parent.Header().Time, block, statedb)
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply eth/state_accessor.go:220]: added `StateAtNextBlock`.

File: eth/state_accessor.go [hunk 5]
```diff
@@ -211,10 +211,11 @@ func (eth *Ethereum) stateAtTransaction(ctx context.Context, block *types.Block,
 	}
 	// Lookup the statedb of parent block from the live database,
 	// otherwise regenerate it on the flight.
-	statedb, release, err := eth.StateAtBlock(ctx, parent, reexec, nil, true, false)
+	statedb, release, err := eth.StateAtNextBlock(ctx, parent, block, reexec, nil, true, false)
 	if err != nil {
 		return nil, vm.BlockContext{}, nil, nil, err
 	}
+
```
Comment (darioush) [inline eth/state_accessor.go:218]: nit: remove wspace

File: eth/tracers/api_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (darioush) [reply eth/tracers/api_test.go:1047]: We already have a lot of tests that we added, so I can approve this PR as is.
My preference is to switch to an additive only model (separate files) as we go so it's more clear what we've added

**Review by darioush — APPROVED** [review]

**Review by darioush — APPROVED** [review]

**Pull request merged**